### ISSUE TITLE
JSONP?

### ISSUE THREAD

**Issue by louh** [pelias/api#193]
Over at [leaflet-geocoder](https://github.com/pelias/leaflet-geocoder) land I am starting to think through whether & how to cover all of the browsers that Leaflet supports, which includes (!!) IE7. What makes life hard for supporting old IE, is that IE9 and under do not support CORS. So people invented the JSONP workaround, which returns cross-domain JSON as a JavaScript in a script tag, but the API also needed to support returning data in that format.

Here's a good summary (the conclusion is that they needed JSONP support): https://blog.algolia.com/jsonp-still-mandatory/

There are a lot of drawbacks to JSONP (not cacheable; can't do POST requests). Suppose we only support CORS and no JSONP. In IE 8 and 9, there is another way to workaround lack of CORS support through its native [XDomainRequest](http://blogs.msdn.com/b/ieinternals/archive/2010/05/13/xdomainrequest-restrictions-limitations-and-workarounds.aspx) object, so I think that's doable (assuming that the API does not rely on headers, like for authentication). Which leaves IE7 in the dust.

So, in summary:
If Pelias supported JSONP, we can support IE7 and be one step closer to supporting all of Leaflet's browser support.
If not, we would have to explicitly drop support for IE7 in the plugin, but that would also make my life a little bit easier, since cross-domain requests are not the only thing standing in the way between us and full compatibility.

**Comment by missinglink** [comment]
@riordan this ^

hey @louh, Dave and I actually had a fairly in-depth discussion about this when I was last in NYC and the team decided to drop JSONP in favour of CORS, although this is always up for discussion.

the rationale behind this was:
- it's massive massive hack :facepalm:
- it doesn't support HTTP, (no POST, PUT and all non-200 codes are not parsed and fail silently)
- it's an attack vector, off the top of my head I can't think of the attack but it's a reason for someone to try to gain control of the server, resulting in security issues and QoS issues

so in an ideal world we wouldn't enable JSONP in our production service (although we would include the [required code](https://github.com/pelias/api/blob/master/middleware/jsonp.js) to enable it in standalone builds)

from an API design standpoint the issue comes in that we would like to be RESTful; not because REST is hipster-cool, but because REST *is* HTTP.

If we decide to support JSONP then we will need to do a couple things:
- move status codes to the JSON body and always return status:200
- allow a 'methodOverride' parameter which says `?method=PUT`

so, in my mind the cons outweigh the pros, although I am totally up for supporting it if it means a fairly large audience of users are now able to use the service who could not before.

two external solutions come to mind, 1) using Flash (or some less :facepalm: polyfill) or 2) using a server-side proxy

**Comment by riordan** [comment]
When it comes to deprecating JSONP in favor of CORS, we'll need to ensure the code for supporting JSONP is available until the complete shutdown.

**Comment by missinglink** [comment]
see https://github.com/pelias/pelias-doc/blob/master/use-cors.md for more info
